File: ts/templatemethod/EmailClient.ts
abstract class EmailClient {
  public sendEmail(recipient: string, subject: string, body: string): void {
    this.connectToServer();
    this.authenticateUser();
    this.specifyRecipients(recipient);
    this.specifySubject(subject);
    this.specifyBody(body);
    this.send();
    this.disconnectFromServer();
  }

  protected abstract connectToServer(): void;
  protected abstract authenticateUser(): void;
  protected abstract specifyRecipients(recipient: string): void;
  protected abstract specifySubject(subject: string): void;
  protected abstract specifyBody(body: string): void;
  protected abstract send(): void;
  protected abstract disconnectFromServer(): void;
}

class GmailEmailClient extends EmailClient {
  protected connectToServer(): void {
    // Connect to Gmail server
  }

  protected authenticateUser(): void {
    // Authenticate user with Gmail
  }

  protected specifyRecipients(recipient: string): void {
    // Specify email recipient(s) for Gmail
  }

  protected specifySubject(subject: string): void {
    // Specify email subject for Gmail
  }

  protected specifyBody(body: string): void {
    // Specify email body for Gmail
  }

  protected send(): void {
    // Send email through Gmail
  }

  protected disconnectFromServer(): void {
    // Disconnect from Gmail server
  }
}

class OutlookEmailClient extends EmailClient {
  protected connectToServer(): void {
    // Connect to Outlook server
  }

  protected authenticateUser(): void {
    // Authenticate user with Outlook
  }

  protected specifyRecipients(recipient: string): void {
    // Specify email recipient(s) for Outlook
  }

  protected specifySubject(subject: string): void {
    // Specify email subject for Outlook
  }

  protected specifyBody(body: string): void {
    // Specify email body for Outlook
  }

  protected send(): void {
    // Send email through Outlook
  }

  protected disconnectFromServer(): void {
    // Disconnect from Outlook server
  }
}

File: java/src/main/java/com/epam/eng/pattern/templatemethod/EmailClient.java
package com.epam.eng.pattern.templatemethod;

public class EmailClient {
    private GmailEmail gmailEmail;
    private OutlookEmail outlookEmail;

    public void sendEmail(String recipient, String subject, String body) {
        connectToServer();
        authenticateUser();
        specifyRecipients(recipient);
        specifySubject(subject);
        specifyBody(body);
        send();
        disconnectFromServer();
    }

    private void connectToServer() {
        // Connect to email server
        if (gmailEmail != null) {
            gmailEmail.connectToServer();
        } else if (outlookEmail != null) {
            outlookEmail.connectToServer();
        }
    }

    private void authenticateUser() {
        // Authenticate user
    }

    private void specifyRecipients(String recipient) {
        // Specify email recipient(s)
    }

    private void specifySubject(String subject) {
        // Specify email subject
    }

    private void specifyBody(String body) {
        // Specify email body
    }

    private void send() {
        // Send email
        if (gmailEmail != null) {
            gmailEmail.send();
        } else if (outlookEmail != null) {
            outlookEmail.send();
        }
    }

    private void disconnectFromServer() {
        // Disconnect from email server
    }
}

class GmailEmail {
    public void connectToServer() {
    }

    public void send() {
    }

    // other methods...
}

class OutlookEmail {
    public void connectToServer() {
    }

    public void send() {
    }

    // other methods...
}

// todo: consider to abstract EmailClient template, so that Gmail and Outlook can implement it
